pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,162 @@
+package namespace
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/apis/cluster/v1alpha1"
+	kerrors "github.com/kcp-dev/kcp/pkg/util/errors"
+)
+
+const clusterLabel = "kcp.dev/cluster"
+
+// reconcileResource is responsible for setting the cluster for a resource of
+// any type, to match the cluster where it's namespace is assigned.
+func (c *Controller) reconcileResource(ctx context.Context, lclusterName string, unstr *unstructured.Unstructured, gvr *schema.GroupVersionResource) error {
+	klog.Infof("Reconciling %s %s/%s", gvr.String(), unstr.GetNamespace(), unstr.GetName())
+
+	// If the resource is not namespaced (incl if the resource is itself a
+	// namespace), ignore it.
+	if unstr.GetNamespace() == "" {
+		klog.Infof("%s %s had no namespace; ignoring", gvr.String(), unstr.GetName())
+		return nil
+	}
+
+	// Align the resource's assigned cluster with the namespace's assigned
+	// cluster.
+	// First, get the namespace object (from the cached lister).
+	ns, err := c.namespaceLister.Get(clusters.ToClusterAwareKey(lclusterName, unstr.GetNamespace()))
+	if err != nil {
+		return kerrors.NewRetryableError(err)
+	}
+
+	lbls := unstr.GetLabels()
+	if lbls == nil {
+		lbls = map[string]string{}
+	}
+
+	old, new := lbls[clusterLabel], ns.Labels[clusterLabel]
+	if old == new {
+		// Already assigned to the right cluster.
+		return nil
+	}
+
+	// Update the resource's assignment.
+	klog.Infof("Patching to update cluster assignment for %s %s/%s: %q -> %q", gvr, ns.Name, unstr.GetName(), old, new)
+	if _, err := c.dynClient.Resource(*gvr).Namespace(ns.Name).Patch(ctx, unstr.GetName(), types.MergePatchType, clusterLabelPatchBytes(new), metav1.PatchOptions{}); err != nil {
+		return kerrors.NewRetryableError(err)
+	}
+	return nil
+}
+
+// reconcileNamespace is responsible for assigning a namespace to a cluster, if
+// it does not already have one.
+//
+// After assigning (or if it's already assigned), this also updates all
+// resources in the namespace to be assigned to the namespace's cluster.
+func (c *Controller) reconcileNamespace(ctx context.Context, lclusterName string, ns *corev1.Namespace) error {
+	klog.Infof("Reconciling Namespace %s", ns.Name)
+
+	if ns.Labels == nil {
+		ns.Labels = map[string]string{}
+	}
+
+	clusterName := ns.Labels[clusterLabel]
+	if clusterName == "" {
+		// Namespace is not assigned to a cluster; assign one.
+		// First, list all clusters.
+		cls, err := c.clusterLister.List(labels.Everything())
+		if err != nil {
+			return kerrors.NewRetryableError(err)
+		}
+
+		// TODO: Filter out un-Ready clusters so a namespace doesn't
```

Comment:
Label selector on your List() call?

Reply:
Do we also apply a label on ready clusters? This might be a good followup.